docs/04-reference/winglang-spec.md

```diff
@@ -782,30 +782,153 @@ let f = (arg1: num, var arg2: num) => {
+#### 1.7.1 Definition
 
-> ```TS
-> let x? = 44;
-> let y = x ?? 55;
-> ```
+The symbol `?` can mark a variable, function argument, return type or a field as optional (same as
+`Option<T>` in Rust).
 
-<details><summary>Equivalent TypeScript Code</summary>
+```js
+struct Person {
+  name: str;
+  address: str?;
+}
 
-> ```TS
-> const x: number? = 44;
-> const y: number = x ?? 55;
+let x = (foo: str?): num? => {
+  print(foo ?? "foo is not defined");
+  return 12;
+};
+
+class Foo {
+  my_opt: num?;
+}
+```
+
+In the above struct, the `address` field, the `foo` argument and the `my_opt` field are all marked
+as optionals using `?`.
+
+This means, for example, that when a `Person` can be initialized without defining the `address`
+field:
+
+```js
+let my_person = Person { name: "david" };
+```
+
+Let's look at various use cases for how to interact with optionals.
+
+#### 1.7.2 Assignment
+
+A reassignable variable or field can only be assigned to a concrete value. There is no way to
+explicitly assign a "null" value:
+
+```js
+let var x: str?;
+x = "hello"; // OK
+
+x = nil;
+//  ^^^ ERROR: no explicit `nil` in this language.
+```
+
+> You can be a smart-ass and do this:
+> ```js
+> let make_nil = (): str? => { return ?; }
+> x = make_nil();
 > ```
+> Good luck with that.
 
-</details>
+#### 1.7.2 Testing if an optional has a value
+
+Object of type `str?` cannot be used within `==` or `!=` expressions because we don't have a null
+value. Instead, use the the `?` operator to get a `bool` that indicates if the optional is defined
+or not:
+
+```js
+let is_address_defined = my_person.address?; // type is `bool`
+
+// or within a condition
+if my_person.address? {
+  print("address is defined but i do not care what it is");
+}
+
+// can be negated
+if !my_person.address? {
+  print("address is not defined");
+}
+```
+
+#### 1.7.3 Unwrapping within conditions
+
+The `if let` statement can be used to test if an optional is defined an *unwrap* it into a
+non-optional variable defined inside the block:
+
+```js
+if let address = my_person.address {
+  assert(address.len > 0);
+  print(address); // address is type `str`
+}
+```
+
+#### 1.7.4 Unwrap and provide a default value
+
+The `??` operator can be used to unwrap or provide a default value:
+
+```js
+let address = my_person.address ?? "Planet Earth";
+```
+
+#### 1.7.5 Throwing an error if a value is not defined (P2)
+
+Another common use case is to bail out with an error in case a value is not defined. To support this
+`throw()` can be used as an r-value. This promotes defensive and safe programming.
+
+```js
+let address = my_person.address ?? throw("address is required");
+```
+
+#### 1.7.6 Optional return types
+
+If a function returns an optional type, use the `return ?` statement to indicate that the value is
+not defined.
+
+```js
+let parse_last_name = (full_name: str): str? => {
+  let parts = full_name.split(" ");
+  if parts.len < 2 {
+    return ?;
+  }
+
+  return parts.at(1);
```

Comment:
There might be a reason to have a version of `at()` that returns `T?` for `Array<T>`, but I understand what's the motivation behind the example 🙂.

Reply:
Yeah, `try_at()` is a great name for that.

Reply:
https://github.com/winglang/wing/issues/1868